src/error/rust.rs

```diff
@@ -1,56 +1,71 @@
 use errno::{set_errno, Errno};
-use std::fmt;
+use std::fmt::Display;
 
 use cosmwasm_vm::VmError;
-use snafu::{ResultExt, Snafu};
+use snafu::Snafu;
 
 use crate::memory::Buffer;
 
 #[derive(Debug, Snafu)]
-#[snafu(visibility = "pub")]
 pub enum Error {
-    #[snafu(display("Wasm Error: {}", source))]
-    WasmErr {
-        source: VmError,
+    #[snafu(display("Null/Empty argument: {}", name))]
+    EmptyArg {
+        name: String,
         #[cfg(feature = "backtraces")]
         backtrace: snafu::Backtrace,
     },
-    #[snafu(display("Caught Panic"))]
-    Panic {
-        #[cfg(feature = "backtraces")]
+    /// Whenever UTF-8 bytes cannot be decoded into a unicode string, e.g. in String::from_utf8 or str::from_utf8.
+    #[snafu(display("Cannot decode UTF8 bytes into string: {}", msg))]
+    InvalidUtf8 {
+        msg: String,
         backtrace: snafu::Backtrace,
     },
-    #[snafu(display("Null/Empty argument passed: {}", name))]
-    EmptyArg {
-        name: &'static str,
+    #[snafu(display("Ran out of gas"))]
+    OutOfGas {
         #[cfg(feature = "backtraces")]
         backtrace: snafu::Backtrace,
     },
-    #[snafu(display("Invalid string format: {}", source))]
-    Utf8Err {
-        source: std::str::Utf8Error,
+    #[snafu(display("Caught Panic"))]
+    Panic {
         #[cfg(feature = "backtraces")]
         backtrace: snafu::Backtrace,
     },
-    #[snafu(display("Ran out of gas"))]
-    OutOfGas {
+    #[snafu(display("Error calling the VM: {}", msg))]
+    VmErr {
+        msg: String,
         #[cfg(feature = "backtraces")]
         backtrace: snafu::Backtrace,
     },
 }
 
-// FIXME: simplify this (and more) when refactoring the errors
-impl From<VmError> for Error {
-    fn from(source: VmError) -> Self {
-        let r: Result<(), VmError> = Err(source);
-        r.context(WasmErr {}).unwrap_err()
+impl Error {
```

Comment:
Nice trial of this new design.

Reply:
Yeah, definitely better than all those free functions, especially since you reduce the imports. And you get free namespacing.

🏅 @reuvenpo